### PR TITLE
[WIP] Redis cluster support

### PR DESCRIPTION
**Pull request by shudipta**
- Redis Clustering
- Add tests for clustering

### CODE REVIEW

File: hack/deploy/setup.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ export MINIKUBE=0
 export MINIKUBE_RUN=0
```
Comment (tamalsaha) [inline hack/deploy/setup.sh]: Fix / rebase

File: hack/docker/redis/4.0/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
+FROM redis
```
Comment (tamalsaha) [inline hack/docker/redis/4.0/Dockerfile]: - Build inside PATCH version
- Upgrade to 4.0.11
- Use alpine base
- Use official image and add trib

File: pkg/admission/validator.go [hunk 3]
```diff
@@ -125,8 +125,23 @@ func ValidateRedis(client kubernetes.Interface, extClient kubedbv1alpha1.KubedbV
 		return err
 	}
 
-	if redis.Spec.Replicas == nil || *redis.Spec.Replicas != 1 {
-		return fmt.Errorf(`spec.replicas "%v" invalid. Value must be one`, redis.Spec.Replicas)
+	if redis.Spec.Mode != api.RedisModeStandalone && redis.Spec.Mode != api.RedisModeCluster {
```
Comment (tamalsaha) [inline pkg/admission/validator.go]: Remove this check.

File: pkg/controller/redis_config.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+	if cm != nil && (cm.Labels[api.LabelDatabaseKind] != api.ResourceKindRedis ||
+		cm.Labels[api.LabelDatabaseName] != redis.Name) {
+		return nil, fmt.Errorf(`intended configmap "%v" already exists`, configmapName)
+	}
+
+	return cm, nil
+}
+
+func (c *Controller) createConfigMap(name string, data map[string]string, redis *api.Redis) (*core.ConfigMap, error) {
```
Comment (tamalsaha) [inline pkg/controller/redis_config.go]: recheck

File: pkg/controller/service.go [hunk 5]
```diff
@@ -163,3 +165,42 @@ func (c *Controller) ensureStatsService(redis *api.Redis) (kutil.VerbType, error
 	}
 	return vt, nil
 }
+
+func (c *Controller) createRedisGoverningService(redis *api.Redis) (string, error) {
```
Comment (tamalsaha) [inline pkg/controller/service.go]: remove?

File: pkg/controller/statefulset.go [hunk 6]
```diff
@@ -239,30 +282,47 @@ func upsertCustomConfig(statefulSet *apps.StatefulSet, redis *api.Redis) *apps.S
...
 
-				// send custom config file path as argument
-				configPath := filepath.Join(CONFIG_MOUNT_PATH, "redis.conf")
-				args := statefulSet.Spec.Template.Spec.Containers[i].Args
-				if len(args) == 0 || args[len(args)-1] != configPath {
-					args = append(args, configPath)
-				}
-				statefulSet.Spec.Template.Spec.Containers[i].Args = args
+				fixIP := filepath.Join(fixIpMountPath, RedisFixIPScriptRelativePath)
+				statefulSet.Spec.Template.Spec.Containers[i].Command = []string{"sh", "-c", fixIP}
```
Comment (tamalsaha) [inline pkg/controller/statefulset.go]: Always or cluster only?

File: pkg/admission/validator.go [hunk 7]
```diff
@@ -125,8 +125,23 @@ func ValidateRedis(client kubernetes.Interface, extClient kubedbv1alpha1.KubedbV
...
+		return fmt.Errorf(`spec.mode "%v" invalid. Value must be one of "%v" or "%v"`,
+			redis.Spec.Mode, api.RedisModeStandalone, api.RedisModeCluster)
+	}
+
+	if redis.Spec.Mode == api.RedisModeCluster && redis.Spec.Cluster == nil {
+		return fmt.Errorf(`spec.cluster "%v" invalid. It is required.`, redis.Spec.Cluster)
+	}
+
+	if redis.Spec.Mode == api.RedisModeCluster &&
```
Comment (tamalsaha) [inline pkg/admission/validator.go]: Block updating cluster Master/ Replicas.

**Comment by shudipta** [comment]
fixes are now in https://github.com/kubedb/redis/pull/94

**Comment by shudipta** [comment]
This pr is no more required since the whole work for cluster support is now on https://github.com/kubedb/redis/pull/94.

**Pull request closed without merging**